Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp

```diff
@@ -217,19 +250,30 @@ static RefPtr<SharedBuffer> expandNativeImageToData(NativeImage& image, ASCIILit
 
 RefPtr<SharedBuffer> createIconDataFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths)
 {
+    // Supported ICO image sizes by ImageIO.
+    constexpr std::array<unsigned, 5> availableLengths { { 16, 32, 48, 128, 256 } };
+    constexpr auto ico = "com.microsoft.ico"_s;
+    auto targetLengths = lengths.empty() ? std::span { availableLengths } : lengths;
+
+    if (RefPtr svgImage = SVGImage::createFromData(data))
+        return expandSVGImageToData(*svgImage, ico, targetLengths);
+
     RefPtr nativeImage = createNativeImageFromData(data, std::nullopt);
     if (!nativeImage)
         return { };
 
-    // Supported ICO image sizes by ImageIO.
-    std::array<unsigned, 5> availableLengths { { 16, 32, 48, 128, 256 } };
-    auto targetLengths = lengths.empty() ? std::span { availableLengths } : lengths;
-    return expandNativeImageToData(*nativeImage, "com.microsoft.ico"_s, targetLengths);
+    return expandNativeImageToData(*nativeImage, ico, targetLengths);
 }
 
+// FIXME: This does not implement preferredSize for SVG at the moment as there are no callers that pass preferredSize.
 RefPtr<ShareableBitmap> decodeImageWithSize(std::span<const uint8_t> data, std::optional<FloatSize> preferredSize)
 {
-    auto nativeImage = createNativeImageFromData(data, preferredSize);
+    RefPtr<NativeImage> nativeImage;
+    if (RefPtr svgImage = SVGImage::createFromData(data))
+        nativeImage = svgImage->nativeImage(svgImage->size());
+    else
+        nativeImage = createNativeImageFromData(data, preferredSize);
```

Comment:
I would reverse the order here. Trying to decode a malformed `BitmapImage` data should return faster than trying to build an `SVGImage` from a malformed `SVGImage` data . This will require creating `Parser`, `Document`, `Page`, `ResourceLoader`, etc.

Reply:
`createNativeImageFromData()` has become a confusing name with this change. 
What does it do exactly?  Does it create a `NativeImage` from any data? I do not think so.

Should it be renamed `tryCreateNativeImageFromBitmapImageData()`?
Should we add a similar one for the `SVGImage`: `tryCreateNativeImageFromSVGImageData()`?

I think adding `try` as a prefix in the name is important since we do not know whether the data is `BitmapImage` or `SVGImage` data or even something else.